src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java

```diff
@@ -0,0 +1,101 @@
+import org.openmrs.reference.page.AdmitToInpatientPage;
+import org.openmrs.reference.page.PatientVisitsDashboardPage;
+import org.openmrs.uitestframework.test.TestData;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+public class EncountersSteps extends Steps {
+
+	private TestData.PatientInfo testPatient;
+
+	@Before(RunTest.HOOK.SELENIUM_ENCOUNTER)
+	public void visitHomePage() {
+		testPatient = createTestPatient();
+		initiateWithLogin();
+	}
+
+	@After(RunTest.HOOK.SELENIUM_ENCOUNTER)
+	public void destroy() {
+		deletePatient(testPatient);
+		quit();
+	}
+
+	@Given("a user clicks on Find Patient app from Home page")
+	public void goToFindPatientRecordPage() {
+		findPatientPage = homePage.goToFindPatientRecord();
+	}
+
+	@Then("the system loads find patient record page")
+	public void systemloadsFindPatientRecordPage() {
+		assertTrue(textExists("Find Patient Record"));
+	}
+
+	@When("a user searches an existing patient and starts the visit")
+	public void searchExistingPatient() {
+		findPatientPage.enterPatient(testPatient.identifier);
+		findPatientPage.waitForPageToLoad();
+		visitsDashboardPage = findPatientPage.clickOnFirstPatient()
+				.startVisit();
+	}
+
+	@Then("the system loads patient visit dashboard page")
+	public void systemloadsVisitDashboardPage() {
+		assertTrue(visitsDashboardPage.containsText("Visits"));
+	}
+
+	// Admission encounter
+	@When("a user clicks on Admit to Inpatient button and selects location")
+	public void goToAdmitToInpatientPage() {
+		visitsDashboardPage.goToAdmitToInpatient().confirm("Inpatient Ward");
+	}
+
+	@And("the system confirms admission encounter is made")
+	public void confirmAdmissionEncounter() {
+		assertTrue(visitsDashboardPage.containsText("Admission"));
+		assertThat(visitsDashboardPage.getEncountersCount(), is(1));
+	}
+
+	// Transfer encounter
+	@When("a user clicks on Trasfer to Ward service button and selects location")
+	public void goToTrasferToWardServicePage() {
+		visitsDashboardPage.goToTransferToWardServicePage().confirm(
+				"Isolation Ward");
+	}
+
+	@Then("the system confirms trasfer within hospital encounter is made")
+	public void confirmTrasferEncounter() {
+		assertTrue(visitsDashboardPage.containsText("Transfer Within Hospital"));
+		assertThat(visitsDashboardPage.getEncountersCount(), is(2));
+	}
+
+	// Discharge encounter
+	@When("a user clicks on Exit from Inpatient button and selects location")
+	public void goToExitFromInpatientPage() {
+		visitsDashboardPage.goToExitFromInpatient().confirm("Isolation Ward");
+	}
+
+	@Then("the system confirms discharge encounter is made")
+	public void confirmDischargeEncounter() throws InterruptedException {
+		assertTrue(visitsDashboardPage.containsText("Discharge"));
+		assertThat(visitsDashboardPage.getEncountersCount(), is(3));
+		visitsDashboardPage.waitForPageToLoad();
+		goToPatientDashboard();
```

Comment:
Can we make this function as a separate step, we might even use it in future in case this tests get too much traffice

Reply:
I thought about the same idea but i realized its not align with the use story for encounter. Actually the user story should end when the 3 encounters have been confirmed at  line line [22](https://github.com/openmrs/openmrs-contrib-qaframework/blob/57bb3fc5467059aa75fb19303cd1e3e5aabf81a5/src/test/resources/features/refapp-2.x/08-encounter/encounters.feature#L22)

From the technical end I thought it very important for the test to return to the clinician dashbaord before it quits the run so that these encounters are observed on the dashbaord. Looking into how the idea would be embedded into the feature file makes it odd man out since its not a user story in the context and I decided to just create a method called [goToPatientDashboard()](https://github.com/openmrs/openmrs-contrib-qaframework/blob/57bb3fc5467059aa75fb19303cd1e3e5aabf81a5/src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java#L97) and invoke it in the last user story [here](https://github.com/openmrs/openmrs-contrib-qaframework/blob/57bb3fc5467059aa75fb19303cd1e3e5aabf81a5/src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java#L94)

Reply:
> I decided to just create a method called [goToPatientDashboard()](https://github.com/openmrs/openmrs-contrib-qaframework/blob/57bb3fc5467059aa75fb19303cd1e3e5aabf81a5/src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java#L97) and invoke it in the last user story [here](https://github.com/openmrs/openmrs-contrib-qaframework/blob/57bb3fc5467059aa75fb19303cd1e3e5aabf81a5/src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java#L94)

I think for better re-using of these methods, we need to extract them , i think you better write another method that handles goToPatientDashboard. this will help us in future to re use this method. this is the same problem that is occuring in [transfertest](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/f2e3df96f7ff815ee8249c4dce61ae460339f770/ui-tests/src/test/java/org/openmrs/reference/page/TransferPatientPage.java#L23) , it will save us time.Thanks

Reply:
> I think for better re-using of these methods, we need to extract them , i think you better write another method that handles goToPatientDashboard. this will help us in future to re use this method

I agree!

To make the method re-susable, I actually created this method earlier on [here](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/f2e3df96f7ff815ee8249c4dce61ae460339f770/ui-tests/src/test/java/org/openmrs/reference/page/PatientVisitsDashboardPage.java#L99) and its just being invoked at code line [100](https://github.com/openmrs/openmrs-contrib-qaframework/blob/d0f633b4f463ba72b6b27d017f738f8513e0aea9/src/test/java/org/openmrs/contrib/qaframework/automation/EncountersSteps.java#L100)